Attempt to add static iptables rules for DOCKER-USER on CentOS 7 (#8357)
docker/docs: network/iptables.md

Docker's rules, add them to the `DOCKER-USER` chain. These rules are loaded
before any rules Docker creates automatically.

### Add a DOCKER-USER filter chain to allow persistent rules 
This can be useful if you need to pre-populate `iptables` rules that need to be in place before 

> Docker runs. The following example creates a new chain named `FILTERS` in which network traffic 

Reviewer: ```suggestion
Docker runs. The following example illustrates how rules can be added to the DOCKER-USER chain
```

> from `INPUT` AND `DOCKER-USER` is put.

Reviewer: ```suggestion

```

Docker's rules, add them to the `DOCKER-USER` chain. These rules are loaded
before any rules Docker creates automatically.

### Add a DOCKER-USER filter chain to allow persistent rules 

> This can be useful if you need to pre-populate `iptables` rules that need to be in place before Docker runs. The following example creates a new chain named `FILTERS` in which network traffic from `INPUT` AND `DOCKER-USER` is put.

Reviewer 2: could you wrap this to ~80 chars, and add whitespace after the heading and before the code block?

```suggestion

This can be useful if you need to pre-populate `iptables` rules that need to be
in place before Docker runs. The following example creates a new chain named
`FILTERS` in which network traffic from `INPUT` AND `DOCKER-USER` is put.

```

Author: Yes - done!

```

> *filter

Reviewer: I'd prefer these
```
*filter

# WAN = ens192 ; LAN = ens160

# Reset counters
:DOCKER-USER - [0:0]

# Flush
-F DOCKER-USER

# Filters :
## Activate established connexions
-A DOCKER-USER -i ens192 -m conntrack --ctstate RELATED,ESTABLISHED -j RETURN

## Allow all on https/http
-A DOCKER-USER -i ens192 -p tcp -m tcp -m conntrack --ctorigdstport 80 -j RETURN
-A DOCKER-USER -i ens192 -p tcp -m tcp -m conntrack --ctorigdstport 443 -j RETURN

## Allow 8080 from ip
-A DOCKER-USER -i ens192 -p tcp -m tcp -m conntrack --ctorigdstport 8080 -s 10.11.11.0/24 -j RETURN
-A DOCKER-USER -i ens192 -p tcp -m tcp -m conntrack --ctorigdstport 8080 -s 10.22.22.0/24 -j RETURN

# Block all external
-A DOCKER-USER -i ens192 -j DROP
-A DOCKER-USER -j RETURN

COMMIT
```

Author: Replaced with your info.

Reviewer: These sets of rules are now redundant(excessive) now that you've mentioned the rules again in the next section (filtering container traffic) :)

Reviewer 3: @arkodg I am following up on this so we can get it into the public docs. Has this issue been resolved or do I still need to remove duplicate information?

:INPUT ACCEPT [0:0]
:FORWARD DROP [0:0]
:OUTPUT ACCEPT [0:0]
:FILTERS - [0:0]
:DOCKER-USER - [0:0]

-F INPUT
-F DOCKER-USER
-F FILTERS

-A INPUT -i lo -j ACCEPT
-A INPUT -p icmp --icmp-type any -j ACCEPT
-A INPUT -j FILTERS

-A DOCKER-USER -i ens33 -j FILTERS

-A FILTERS -m state --state ESTABLISHED,RELATED -j ACCEPT
-A FILTERS -m state --state NEW -s 1.2.3.4/32 -j ACCEPT
-A FILTERS -m state --state NEW -m tcp -p tcp --dport 22 -j ACCEPT
-A FILTERS -m state --state NEW -m tcp -p tcp --dport 23 -j ACCEPT
-A FILTERS -m state --state NEW -m tcp -p tcp --dport 80 -j ACCEPT
-A FILTERS -m state --state NEW -m tcp -p tcp --dport 443 -j ACCEPT
-A FILTERS -j REJECT --reject-with icmp-host-prohibited

COMMIT
```

Load this into the kernel with:

> ```

Reviewer 2: Perhaps add whitespace here as well (had some occurrences where rendering had issues if missing), probably add a `bash` code-hint, and a prompt (`$`)

```suggestion

```bash
$ iptables-restore -n /etc/iptables.conf
```

Author: Added.

iptables-restore -n /etc/iptables.conf
```

> Use the previous FILTERS chain setup with the following configuration to allow `icmp` to the docker host and allow host port 22 access and container port 5222 access:

Reviewer 2: Could you wrap this, and add backticks on `FILTERS` (likely `22` and `5222` as well);

```suggestion
Use the previous `FILTERS` chain setup with the following configuration to allow
`icmp` to the docker host and allow host port `22` access and container port `5222`
access:
```

Author: Got 'em.

```
-A FILTERS -p icmp --icmp-type any -s client_a/32 -j ACCEPT
-A FILTERS -p icmp --icmp-type any -j DROP
-A FILTERS -m conntrack --ctstate ESTABLISHED,RELATED -j ACCEPT
-A FILTERS -m conntrack --ctstate NEW -m tcp -p tcp -s client_a/32 --dport 22 -j ACCEPT
-A FILTERS -m conntrack --ctstate NEW -m tcp -p tcp -s client_a/32 --ctorigdstport 5222 -j ACCEPT
-A FILTERS -j DROP
```

> **Note**: `--ctorigdstport` matches the destination port on the packet that initiated the connection, not the destination port on the packet being filtered. Therefore, responses to requests from Docker to other servers have `SPT=80` and match `--ctorigdstport 80`.

Reviewer 2: Could you format this as a note? (`> **Note**:`), not sure, but should there be an oxford-comma before "and match"?

```suggestion
> **Note**: `--ctorigdstport` matches the destination port on the packet that
> initiated the connection, not the destination port on the packet being filtered.
> Therefore, responses to requests from Docker to other servers have `SPT=80`,
> and match `--ctorigdstport 80`.
```

Author: Got 'em.

> For tighter control, all rules allowing the connection should have `--ctdir` added to specifically express their meaning, as shown in the following example:

Reviewer 2: ```suggestion
For tighter control, all rules allowing the connection should have `--ctdir` added
to specifically express their meaning, as shown in the following example:
```

Author: Got it.

topic. See the [Netfilter.org HOWTO](https://www.netfilter.org/documentation/HOWTO/NAT-HOWTO.html)
for a lot more information.

> ### Filtering container and host traffic

Reviewer: ```suggestion
### Filtering container traffic
```

## Next steps

- Read [Docker Reference Architecture: Designing Scalable, Portable Docker Container Networks]

> (https://success.docker.com/Architecture/Docker_Reference_Architecture%3A_Designing_Scalable%2C_Portable_Docker_Container_Networks)

Reviewer 2: looks like the added whitespace broke the link

To wrap links, the wrapping needs to be done in the links' "caption";

```Markdown
- Read [Docker Reference Architecture: Designing Scalable, Portable Docker
  Container Networks](https://success.docker.com/Architecture/Docker_Reference_Architecture%3A_Designing_Scalable%2C_Portable_Docker_Container_Networks)
```

That's a bit awkward sometimes, so an alternative could be to rephrase it slightly;

```Markdown
- Read "Docker Reference Architecture: Designing Scalable, Portable Docker
  Container Networks" on [success.docker.com](https://success.docker.com/Architecture/Docker_Reference_Architecture%3A_Designing_Scalable%2C_Portable_Docker_Container_Networks)
```

Or just don't wrap it (which I think in this case would be ok as an exception)

topic. See the [Netfilter.org HOWTO](https://www.netfilter.org/documentation/HOWTO/NAT-HOWTO.html)
for a lot more information.

### Filtering container and host traffic
The following example provides a set of filters and uses those filters for container and host traffic: 

#### To filter container traffic:

```
*filter

# WAN = eth0 ; LAN = eth1

# Reset counters
:DOCKER-USER - [0:0]

# Flush
-F DOCKER-USER

# Filters :
## Activate established connexions
-A DOCKER-USER -i eth0 -m conntrack --ctstate RELATED,ESTABLISHED -j RETURN

## Allow all on https/http
-A DOCKER-USER -i eth0 -p tcp -m tcp -m conntrack --ctorigdstport 80 -j RETURN
-A DOCKER-USER -i eth0 -p tcp -m tcp -m conntrack --ctorigdstport 443 -j RETURN

## Allow 8080 from ip
-A DOCKER-USER -i eth0 -p tcp -m tcp -m conntrack --ctorigdstport 8080 -s 10.11.11.0/24 -j RETURN
-A DOCKER-USER -i eth0 -p tcp -m tcp -m conntrack --ctorigdstport 8080 -s 10.22.22.0/24 -j RETURN

# Block all external
-A DOCKER-USER -i eth0 -j DROP
-A DOCKER-USER -j RETURN

COMMIT
```

> 

Reviewer: ```suggestion
> **Note**: `--ctorigdstport` matches the destination port on the packet that initiated the connection, 
	not the destination port on the packet being filtered. Therefore, responses to requests from Docker 
	to other servers have `SPT=80`, and match `--ctorigdstport 80`.
	
	For tighter control, all rules allowing the connection should have `--ctdir` added to specifically 
	express their meaning, as shown in the following example:
```

#### To filter host traffic:

> **Note**: Set the filter for WAN based on your host WAN interface.

```
*filter

> # WAN = ens0 ; LAN = eth0

Reviewer: ```suggestion
# WAN = eth0 ; LAN = eth1
```
would be a better example

:FORWARD DROP [0:0]
:OUTPUT ACCEPT [0:0]
:FILTERS - [0:0]
:FILTERS-LAN - [0:0]

# Flush
-F INPUT
-F FILTERS
-F FILTERS-LAN

# Select
-A INPUT -i lo -j ACCEPT
-A INPUT -i eth1 -j FILTERS-LAN
-A INPUT -i eth0 -j FILTERS

# Filters
## Activate established connexions
-A FILTERS -m state --state ESTABLISHED,RELATED -j ACCEPT

## Monitoring
-A FILTERS -s 10.1.1.1/32 -p udp -m udp --dport 161 -j ACCEPT
-A FILTERS -s 10.1.1.1/32 -p tcp -m tcp --dport 5666 -j ACCEPT
-A FILTERS -s 10.1.1.1/32 -p icmp --icmp-type any -j ACCEPT

## Admin ssh
-A FILTERS -s 10.0.0.1/32 -p tcp -m tcp --dport 22 -j ACCEPT
-A FILTERS -s 10.0.1.1/32 -p tcp -m tcp --dport 22 -j ACCEPT

## Admin ping
-A FILTERS -s 10.0.0.1/32 -p icmp --icmp-type any -j ACCEPT
-A FILTERS -s 10.0.1.1/32 -p icmp --icmp-type any -j ACCEPT

## Drop public in
-A FILTERS -j DROP

# Filters-LAN
## Activate established connexions
-A FILTERS-LAN -m state --state ESTABLISHED,RELATED -j ACCEPT

## Admin allow all
-A FILTERS-LAN -s 10.0.1.1/32 -j ACCEPT

## Ping
-A FILTERS-LAN -s 10.0.1.1/24 -p icmp --icmp-type any -j ACCEPT

## Log and Drop lan in
-A FILTERS-LAN -j LOG --log-prefix "[LAN BLOCK] "
-A FILTERS-LAN -j DROP

## Commit
COMMIT

> ```

Reviewer: Load this into the kernel with:
	
	```bash
	$ iptables-restore -n /etc/iptables.conf
	```

-A FILTERS -s 10.1.1.1/32 -p icmp --icmp-type any -j ACCEPT

## Admin ssh
-A FILTERS -s 10.0.0.1/32 -p tcp -m tcp --dport 22 -j ACCEPT
-A FILTERS -s 10.0.1.1/32 -p tcp -m tcp --dport 22 -j ACCEPT

## Admin ping
-A FILTERS -s 10.0.0.1/32 -p icmp --icmp-type any -j ACCEPT
-A FILTERS -s 10.0.1.1/32 -p icmp --icmp-type any -j ACCEPT

## Drop public in
-A FILTERS -j DROP
```

#### To filter container traffic:

```
*filter

# WAN = ens192 ; LAN = ens160

# Reset counters
:DOCKER-USER - [0:0]

# Flush
-F DOCKER-USER

# Filters :
## Activate established connexions
-A DOCKER-USER -i ens192 -m conntrack --ctstate RELATED,ESTABLISHED -j RETURN

## Allow all on https/http
-A DOCKER-USER -i ens192 -p tcp -m tcp -m conntrack --ctorigdstport 80 -j RETURN
-A DOCKER-USER -i ens192 -p tcp -m tcp -m conntrack --ctorigdstport 443 -j RETURN

## Allow 8080 from ip
-A DOCKER-USER -i ens192 -p tcp -m tcp -m conntrack --ctorigdstport 8080 -s 10.11.11.0/24 -j RETURN
-A DOCKER-USER -i ens192 -p tcp -m tcp -m conntrack --ctorigdstport 8080 -s 10.22.22.0/24 -j RETURN

# Block all external
-A DOCKER-USER -i ens192 -j DROP
-A DOCKER-USER -j RETURN

COMMIT
```

> #### To filter host traffic:

Reviewer: can you please add a comment to replace the interface with what the user has on their system (ens0, eth0)

Author: OK - see if this is better.

Reviewer: I'd prefer calling it `eth0` and adding a Note for users to set it based on their host WAN interface
